plutus-ledger-constraints/src/Ledger/Constraints/TxConstraints.hs

```diff
@@ -357,15 +361,70 @@ mustPayWithDatumToPubKeyAddress
     -> Datum
     -> Value
     -> TxConstraints i o
-mustPayWithDatumToPubKeyAddress pkh skh datum =
-    singleton . MustPayToPubKeyAddress pkh (Just skh) (Just datum)
+mustPayWithDatumToPubKeyAddress pkh skh datum vl =
+    singleton (MustPayToPubKeyAddress pkh (Just skh) (Just datum) Nothing vl)
+
+{-# INLINABLE mustOutputInlineValidator #-}
+-- | @mustOutputInlineValidator@ is a helper that calls @mustOutputInlineScript@.
+mustOutputInlineValidator
+    :: forall i o
+    . Address
+    -> ValidatorHash
+    -> Maybe Datum
+    -> Value
+    -> TxConstraints i o
+mustOutputInlineValidator addr (ValidatorHash vh) = mustOutputInlineScript addr (ScriptHash vh)
+
+{-# INLINABLE mustOutputInlineMintingPolicy #-}
+-- | @mustOutputInlineMintingPolicy@ is a helper that calls @mustOutputInlineScript@.
+mustOutputInlineMintingPolicy
+    :: forall i o
+    . Address
+    -> MintingPolicyHash
+    -> Maybe Datum
+    -> Value
+    -> TxConstraints i o
+mustOutputInlineMintingPolicy addr (MintingPolicyHash vh) = mustOutputInlineScript addr (ScriptHash vh)
+
+{-# INLINABLE mustOutputInlineScript #-}
+-- | @mustOutputInlineScript addr scriptHash d v@ creates a transaction output
+-- with an inline script. This allows the script to be used as a reference script.
+--
+-- If used in 'Ledger.Constraints.OffChain', this constraint creates an
+-- output with @addr@, @scriptHash@, @d@ and @v@ and maybe adds @d@ in the transaction's
+-- datum witness set.
+--
+-- If used in 'Ledger.Constraints.OnChain', this constraint verifies that @d@ is
+-- part of the datum witness set and that the transaction output with
+-- @addr@, @scriptHash@, @d@ and @v@ is part of the transaction's outputs.
+mustOutputInlineScript
+    :: forall i o
+    . Address
+    -> ScriptHash
+    -> Maybe Datum
+    -> Value
+    -> TxConstraints i o
+mustOutputInlineScript
+    (Address (PubKeyCredential pkh) (Just (StakingHash (PubKeyCredential sh)))) scriptHash datum value =
+        singleton (MustPayToPubKeyAddress (PaymentPubKeyHash pkh) (Just (StakePubKeyHash sh)) datum (Just scriptHash) value)
+mustOutputInlineScript
+    (Address (PubKeyCredential pkh) Nothing) scriptHash datum value =
+        singleton (MustPayToPubKeyAddress (PaymentPubKeyHash pkh) Nothing datum (Just scriptHash) value)
+mustOutputInlineScript
+    (Address (ScriptCredential vh) (Just (StakingHash (ScriptCredential (ValidatorHash sh))))) scriptHash datum value =
+        singleton (MustPayToOtherScript vh (Just (StakeValidatorHash sh)) (fromMaybe unitDatum datum) (Just scriptHash) value)
+mustOutputInlineScript
+    (Address (ScriptCredential vh) Nothing) scriptHash datum value =
+        singleton (MustPayToOtherScript vh Nothing (fromMaybe unitDatum datum) (Just scriptHash) value)
+mustOutputInlineScript
+    addr _ _ _ = Haskell.error $ "Ledger.Constraints.TxConstraints.mustOutputInlineScript: unsupported address " Haskell.++ Haskell.show addr
```

Comment:
Isn't the `Address` pattern matching complete? What data constructor are we missing?